junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/EngineDiscoveryResultValidator.java

```diff
@@ -37,30 +38,39 @@ void validate(TestEngine testEngine, TestDescriptor root) {
 			() -> String.format(
 				"The discover() method for TestEngine with ID '%s' must return a non-null root TestDescriptor.",
 				testEngine.getId()));
-		Preconditions.condition(isAcyclic(root),
-			() -> String.format("The discover() method for TestEngine with ID '%s' returned a cyclic graph.",
-				testEngine.getId()));
+		ArrayList<UniqueId> cyclicUIDs = getCyclicUIDs(root);
+		Preconditions.condition(cyclicUIDs.isEmpty(),
+			() -> String.format("The discover() method for TestEngine with ID '%s' returned a cyclic graph: "
+					+ "Unique ID(s) '%s' enter a cycle",
+				testEngine.getId(), cyclicUIDs));
 	}
 
 	/**
 	 * @return {@code true} if the tree does <em>not</em> contain a cycle; else {@code false}.
 	 */
 	boolean isAcyclic(TestDescriptor root) {
+		return getCyclicUIDs(root).isEmpty();
+	}
+
+	ArrayList<UniqueId> getCyclicUIDs(TestDescriptor root) {
```

Comment:
Consider using an `Optional` instead.

Reply:
sure, totally make sense for current implementation, but again emitting all UIDs is still a tempting alternative

Reply:
fixed now, thanks a lot!